### PR TITLE
metrics: Add k8s client rate limiter latency metric

### PR DESCRIPTION
**Pull request by ysksuzuki**
The rate limiter metrics visualize the extent of delays caused by the k8s client-side rate limit, providing valuable insights for making decisions on configuration adjustments.

~~Also, this commit fixes a conflict with controller-runtime on the metrics registration.~~

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

### CODE REVIEW

**Comment by ysksuzuki** [comment]
Hit https://github.com/cilium/cilium/issues/25524

**Comment by aojea** [comment]
watchers latency is not really important, since they are long lived sessions 

The metrics for client-go when not used for watch can be loaded automatically just importing
```
	_ "k8s.io/component-base/metrics/prometheus/clientgo" // for client metric registration
```

**Comment by ysksuzuki** [comment]
Hi @aojea , Thank you for your comment.

> watchers latency is not really important, since they are long lived sessions

This PR adds the rate limiter latency metric, and I think it's important because we have had a problem before with write operations being delayed by rate limits in our environment. I wanted to know if it's necessary to adjust the QPS configs because the default config in client-go is a bit low.

client-go QPS default is 5
https://github.com/kubernetes/client-go/blob/b46677097d03b964eab2d67ffbb022403996f4d4/rest/config.go#L44

controller-runtime QPS default is 20
https://github.com/kubernetes-sigs/controller-runtime/blob/f6f37e6cc1ec7b7d18a266a6614f86df211b1a0a/pkg/client/config/config.go#L102 

> The metrics for client-go when not used for watch can be loaded automatically just importing

As far as I looked into this, we can't do this because controller-runtime also calls the client-go metrics registration [here](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/metrics/client_go_adapter.go#L52) and subsequent calls will be ignored. [Register](https://github.com/kubernetes/client-go/blob/master/tools/metrics/metrics.go#L124) can only be called once. 

This also calls `register`
https://github.com/kubernetes/component-base/blob/master/metrics/prometheus/restclient/metrics.go#L188

Aso, Cilium has already exported the client-go metrics under the names `cilium_k8s_client_api_latency_time_seconds` and `cilium_k8s_client_api_calls_total` [here](https://github.com/cilium/cilium/blob/main/pkg/metrics/metrics.go#L1184-L1204). We need to announce the breaking change if we use `k8s.io/component-base/metrics/prometheus/clientgo`.

**Comment by aojea** [comment]
> This PR adds the rate limiter latency metric, and I think it's important because we have had a problem before with write operations being delayed by rate limits in our environment. I wanted to know if it's necessary to adjust the QPS configs because the default config in client-go is a bit low.

my point is that this file seems to be used only for watchers, and watchers are not affected by the rate limiter,

https://github.com/kubernetes/kubernetes/blob/b2522655b3ab70de0a5cc1af0d9ca71e41b25c97/staging/src/k8s.io/client-go/rest/request.go#L702-L705

 but I'm not familiar enough with cilium codebase to know if this client is used for more operations

> As far as I looked into this, we can't do this because controller-runtime also calls the client-go metrics registration [here](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/metrics/client_go_adapter.go#L52) and subsequent calls will be ignored. [Register](https://github.com/kubernetes/client-go/blob/master/tools/metrics/metrics.go#L124) can only be called once.

Showing my ignorance here about cilium codebase :) my point is about encouraging all projects that import client-go to use all the existing client-go metrics, it's an area I was working on the past to improve the "client-side network visibility" and I'm happy to get feedback and new ideas for improvements, as you can see there are some few interesting metrics that can be useful to troubleshoot client networking problems

https://github.com/kubernetes/kubernetes/pull/117295
https://github.com/kubernetes/kubernetes/pull/108296

**Comment by ysksuzuki** [comment]
> my point is that this file seems to be used only for watchers, and watchers are not affected by the rate limiter,

Thank you for clarifying it. This client is used for the write operations too.

cilium_k8s_client_api_latency_time_seconds_bucket

> my point is about encouraging all projects that import client-go to use all the existing client-go metrics

Yeah, it makes sense to import `k8s.io/component-base/metrics/prometheus/clientgo` and use all the existing client-go metrics. Cilium v1.13 started using controller-runtime and seems to have a conflict with the metrics registration. I think we should create an issue on the controller-runtime repo and encourage them to use all the existing client-go metrics instead of registering `RequestResult` only.

https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/metrics/client_go_adapter.go#L52

**Comment by ysksuzuki** [comment]
> I think we should create an issue on the controller-runtime repo and encourage them to use all the existing client-go metrics instead of registering RequestResult only.

contoller-runtime introduced `rest_client_request_duration_seconds`, `rest_client_request_size_bytes` and `rest_client_response_size_bytes`, but reverted because the cardinality of those metrics were too high when CR is used to communicate with a lot of different clusters. Considering this, we should not register all the client-go metrics by default.

https://github.com/kubernetes-sigs/controller-runtime/pull/2298

**Comment by aojea** [comment]
> Considering this, we should not register all the client-go metrics by default.

wow, I didn't expect that, client-go use to talk with apiserver only.
It seems they are reusing the same client for multiple cluster that of course is not something scales well

**Comment by ysksuzuki** [comment]
They seem to intentionally override the host label to keep the cardinality low but in fact, it was too high, then reverted. I'm not sure why it turned out that way.

https://github.com/kubernetes-sigs/controller-runtime/pull/2217
> I think the cardinality should be fine as the host label intentionally only contains the "template url".

**Review by tommyp1ckles — APPROVED** [review]

**Comment by ysksuzuki** [comment]
Resolved the conflict and rebased

**Comment by ysksuzuki** [comment]
Hi @tommyp1ckles , I resolved a conflict and rebased my branch. Please take a look

**Comment by joestringer** [comment]
I don't see any docs updates for this, does it make sense to add some line to `Documentation/observability/metrics.rst` to help users to find these new metrics?

I've added the release-note/minor label to the PR because there are user-facing changes here. I also tried to kick the ginkgo CI into gear (the only remaining "required" test job). If it doesn't run & provide results in the next 20-30 minutes then it might just need the PR to be rebased.

**Comment by ysksuzuki** [comment]
> I don't see any docs updates for this, does it make sense to add some line to Documentation/observability/metrics.rst to help users to find these new metrics?

Yeah, it makes sense. As far as I looked into it, there's no description for the  k8s rest client metrics. So I will add a section for it.

**Review by qmonnet — APPROVED** [review]
Doc change looks good :+1:

**Comment by joestringer** [comment]
> Also, this commit fixes a conflict with controller-runtime on the metrics registration.

Typically we would suggest to propose a specific PR only for the fix, so that we can backport that fix to all releases that are affected. As I understand from our discussion during the community APAC call today, part of this change fixes a metrics bug in 1.13. Would you consider splitting that out and proposing as a dedicated PR so that we can ensure that gets fixed in the relevant releases?

**Review by joestringer — APPROVED** [review]
In principle I think the @cilium/sig-policy codeowner request is only added because of the metrics docs page. I can approve on behalf of that codeowner.

I didn't realize that there was a fix in this PR since it was also adding a new metrics feature. I'd suggest submitting fixes in separate PRs from features, so that we can more easily identify them and review in context of the fix, and then we can also backport those fixes to affected releases.

**Comment by ysksuzuki** [comment]
> Would you consider splitting that out and proposing as a dedicated PR so that we can ensure that gets fixed in the relevant releases?

Sure, I will do that. I should have done that in the first place😅

**Comment by ysksuzuki** [comment]
> Also, this commit fixes a conflict with controller-runtime on the metrics registration.

I tested this change with the latest main and found that the conflict issue doesn't persist. That's because the order in which the init function is called has changed. I couldn't identify which change caused the dependency graph to change, though.

@joestringer It seems that the conflict issue with controller-runtime is not the case anymore when it comes to the latest main. (It still persists with v1.13.4 though)  I think I need to modify this PR. I'm sorry for causing confusion.


latest main: 
`github.com/cilium/cilium/pkg/k8s/watchers` init -> `github.com/cilium/cilium/operator/metrics` -> `sigs.k8s.io/controller-runtime/pkg/metrics` init

v1.13.4:
`github.com/cilium/cilium/operator/metrics` -> `sigs.k8s.io/controller-runtime/pkg/metrics` init -> `github.com/cilium/cilium/pkg/k8s/watchers` init -

**Comment by ysksuzuki** [comment]
I can see the rate limiter metrics
rate-limiter

**Comment by joestringer** [comment]
> It seems that the conflict issue with controller-runtime is not the case anymore when it comes to the latest main. (It still persists with v1.13.4 though) 

OK I see. If it's only a regression in v1.13 now, then one option could be to propose the fix only for the v1.13 branch. That way we can ensure that users can upgrade to v1.13.x releases without breaking those metrics.

**Comment by ysksuzuki** [comment]
>  If it's only a regression in v1.13 now, then one option could be to propose the fix only for the v1.13 branch. That way we can ensure that users can upgrade to v1.13.x releases without breaking those metrics.

I will raise a PR for the v1.13 branch

**Comment by ysksuzuki** [comment]
I have created a fix for v1.13
https://github.com/cilium/cilium/pull/26412

File: pkg/k8s/watchers/watcher.go [hunk 1]
```diff
@@ -93,13 +93,27 @@ func init() {
...
+	// Without calling Register function here, controller-runtime will have
+	// a chance to override our metrics registration.
+	registerOps := k8s_metrics.RegisterOpts{
 		ClientCertExpiry:      nil,
 		ClientCertRotationAge: nil,
-		RequestLatency:        &k8sMetrics{},
-		RateLimiterLatency:    nil,
-		RequestResult:         &k8sMetrics{},
-	})
+		RequestLatency:        &requestLatencyAdapter{},
```
Comment (joestringer) [inline pkg/k8s/watchers/watcher.go:99]: Does it make sense to separate this commit out into its own commit?

I'm still a little bit confused about how this change relates to the new metric, it seems like separate refactoring to address another issue.

But maybe the @cilium/sig-k8s reviewers can chime in here for closer inspection.

Comment (ysksuzuki) [reply pkg/k8s/watchers/watcher.go:99]: OK, ~~I will separate this change into its own commit.~~ Dropped the refactoring 

Although we can export the rate limiter metric without this change of the init function, there is a possibility that the dependency graph may change due to future changes, and the client-go metrics, including the rate limiter metric, may be missing, and it cannot be detected through testing. So I decided to include this change.

Comment (ysksuzuki) [reply pkg/k8s/watchers/watcher.go:99]: >  it seems like separate refactoring to address another issue.

But, yeah, you are right. I think I should tackle the conflict issue(with controller-runtime) on another issue. Let's drop this change and focus on adding new metric here.

**Comment by ysksuzuki** [comment]
@joestringer Can we move this PR forward?

**Review by joestringer — APPROVED** [review]
